src/coreComponents/constitutive/fluid/multifluid/compositional/CompositionalMultiphaseFluidUpdates.hpp

```diff
@@ -0,0 +1,431 @@
+    {
+      compMoleFrac[ic] = composition[ic];
+    }
+  }
+
+  // 2. Compute phase fractions and phase component fractions
+  m_flash.compute( pressure,
+                   temperature,
+                   compMoleFrac.toSliceConst(),
+                   phaseFrac,
+                   phaseCompFrac );
+
+  // 3. Calculate the phase densities
+  m_phase1.density.compute( pressure,
+                            temperature,
+                            phaseCompFrac[0].toSliceConst(),
+                            phaseDens[0],
+                            phaseMassDens[0],
+                            m_useMass );
+  m_phase2.density.compute( pressure,
+                            temperature,
+                            phaseCompFrac[1].toSliceConst(),
+                            phaseDens[1],
+                            phaseMassDens[1],
+                            m_useMass );
+  if constexpr (2 < FLASH::getNumberOfPhases())
+  {
+    m_phase3.density.compute( pressure,
+                              temperature,
+                              phaseCompFrac[2].toSliceConst(),
+                              phaseDens[2],
+                              phaseMassDens[2],
+                              m_useMass );
+  }
+
+  // 4. Calculate the phase viscosities
+  m_phase1.viscosity.compute( pressure,
+                              temperature,
+                              phaseCompFrac[0].toSliceConst(),
+                              phaseVisc[0],
+                              m_useMass );
+  m_phase2.viscosity.compute( pressure,
+                              temperature,
+                              phaseCompFrac[1].toSliceConst(),
+                              phaseVisc[1],
+                              m_useMass );
+  if constexpr (2 < FLASH::getNumberOfPhases())
+  {
+    m_phase3.viscosity.compute( pressure,
+                                temperature,
+                                phaseCompFrac[2].toSliceConst(),
+                                phaseVisc[2],
+                                m_useMass );
+  }
+
+  // 5. if mass variables used instead of molar, perform the conversion
+
+  if( m_useMass )
+  {
+
+    // unfortunately here, we have to copy the molecular weight coming from PVT package...
+    real64 phaseMolecularWeight[maxNumPhase]{};
+    for( integer ip = 0; ip < numPhase; ++ip )
+    {
+      phaseMolecularWeight[ip] = 1.0;
+    }
+
+    // convert mole fractions to mass fractions
+    convertToMassFractions< maxNumComp >( phaseMolecularWeight,
+                                          phaseFrac,
+                                          phaseCompFrac );
+
+  }
+
+  // 6. Compute total fluid mass/molar density
+
+  computeTotalDensity< maxNumComp, maxNumPhase >( phaseFrac,
+                                                  phaseDens,
+                                                  totalDens );
+}
+
+template< typename FLASH, typename PHASE1, typename PHASE2, typename PHASE3 >
```

Comment:
I have never been a fan of this design where we have separate but nearly identical implementations, one with and one without derivatives.  I think this one only gets used in the hydrostatic initialization call?  Would be nice to remove one of them to avoid keeping redundant code aligned.  Also, I think the viscosity call below has no impact on the simulation at all.

Reply:
This looks like it needs to be a concerted effort across the different fluid model types. I will add a task to #2465 to follow up on this.